### PR TITLE
feat(lidar_centerpoint_tvm): add lidar centerpoint tvm package

### PR DESCRIPTION
**Pull request by angry-crab**
Signed-off-by: Xinyu Wang <xinyu.wang@tier4.jp>

## Description

Reopen #1898 because the auther is inactive and the branch cannot be rebased. 
Provide another lidar centerpoint package using the TVM inference engine compared to the original one using TensorRT inference engine. And this package's directory structure is consistent the original package.

## Related links

Related issue #908 

## Tests performed

Tested locally.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [ ] I've confirmed the [contribution guidelines].
- [ ] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by angry-crab** [comment]
FYI, I added my name to the maintainer's list since the author has be inactive for a while.

File: perception/lidar_centerpoint_tvm/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+set(tvm_runtime_DIR ${tvm_vendor_DIR})
+find_package(tvm_runtime CONFIG REQUIRED)
+
+# Gather neural network information.
+set(${PROJECT_NAME}_BACKEND llvm CACHE STRING "${PROJECT_NAME} neural network backend")
+set(MODEL_NAME_ENCODER centerpoint_encoder)
+
+# Get neural network.
+set(NN_DEPENDENCY_ENCOER "")
```
Comment (mitsudome-r) [inline perception/lidar_centerpoint_tvm/CMakeLists.txt:15]:
```suggestion
set(NN_DEPENDENCY_ENCODER "")
```

File: perception/lidar_centerpoint_tvm/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+get_neural_network(${MODEL_NAME_ENCODER} ${${PROJECT_NAME}_BACKEND} NN_DEPENDENCY_ENCOER)
```
Comment (mitsudome-r) [inline perception/lidar_centerpoint_tvm/CMakeLists.txt:16]:
```suggestion
get_neural_network(${MODEL_NAME_ENCODER} ${${PROJECT_NAME}_BACKEND} NN_DEPENDENCY_ENCODER)
```

File: perception/lidar_centerpoint_tvm/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+set(MODEL_NAME_BACKBONE centerpoint_backbone)
+
+# Get neural network.
+set(NN_DEPENDENCY_BACKBONE "")
+get_neural_network(${MODEL_NAME_BACKBONE} ${${PROJECT_NAME}_BACKEND} NN_DEPENDENCY_BACKBONE)
+
+if((NOT NN_DEPENDENCY_ENCOER STREQUAL "") AND (NOT NN_DEPENDENCY_BACKBONE STREQUAL ""))
```
Comment (mitsudome-r) [inline perception/lidar_centerpoint_tvm/CMakeLists.txt:24]:
```suggestion
if((NOT NN_DEPENDENCY_ENCODER STREQUAL "") AND (NOT NN_DEPENDENCY_BACKBONE STREQUAL ""))
```

File: perception/lidar_centerpoint_tvm/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    lib/ros_utils.cpp
+    lib/network/scatter.cpp
+    lib/preprocess/pointcloud_densification.cpp
+    lib/preprocess/voxel_generator.cpp
+    lib/preprocess/generate_features.cpp
+    lib/postprocess/circle_nms.cpp
+    lib/postprocess/generate_detected_boxes.cpp
+  )
+
+  add_dependencies(${PROJECT_NAME} ${NN_DEPENDENCY_ENCOER})
```
Comment (mitsudome-r) [inline perception/lidar_centerpoint_tvm/CMakeLists.txt:40]:
```suggestion
  add_dependencies(${PROJECT_NAME} ${NN_DEPENDENCY_ENCODER})
```

File: perception/lidar_centerpoint_tvm/README.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+| `densification_num_past_frames` | int    | `1`           | the number of past frames to fuse with the current frame    |
+
+### Bounding Box
+
+The lidar segmentation node establishes a bounding box for the detected obstacles.
+The `L-fit` method of fitting a bounding box to a cluster is used for that.
+
+### Limitation and Known Issue
+
+Due to an accurary issue of `centerpoint` model, `vulkan` cannot be used at the moment.
```
Comment (mitsudome-r) [inline perception/lidar_centerpoint_tvm/README.md:54]:
```suggestion
Due to an accuracy issue of `centerpoint` model, `vulkan` cannot be used at the moment.
```

File: perception/lidar_centerpoint_tvm/include/lidar_centerpoint_tvm/centerpoint_config.hpp [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    offset_y_ = range_min_y_ + voxel_size_y_ / 2;
+    offset_z_ = range_min_z_ + voxel_size_z_ / 2;
+    down_grid_size_x_ = grid_size_x_ / downsample_factor_;
+    down_grid_size_y_ = grid_size_y_ / downsample_factor_;
+  };
+
+  // input params
+  std::size_t class_size_{3};
+  const std::size_t point_dim_size_{3};  // x, y and z
+  std::size_t point_feature_size_{4};    // x, y, z and timelag
```
Comment (mitsudome-r) [inline perception/lidar_centerpoint_tvm/include/lidar_centerpoint_tvm/centerpoint_config.hpp:81]:
```suggestion
  std::size_t point_feature_size_{4};    // x, y, z and time lag
```

**Review by mitsudome-r — COMMENTED** [review]
The original PR is approved so we can merged this once some spell-check issues are solved.

**Review by mitsudome-r — APPROVED** [review]

**Pull request merged**